compiler: custom method called on non-identifier leads to panic during compilation

The following test fails to compile:
```
func Test(t *testing.T) {
	src := `package foo
		import "github.com/nspcc-dev/neo-go/pkg/interop"
		import "github.com/nspcc-dev/neo-go/pkg/interop/runtime"

		func Main() bool {
			var a interop.Hash160
			return runtime.GetCallingScriptHash().Equals(a)
		}`
	eval(t, src, false)
}
```
Panic trace:
```
=== RUN   Test
--- FAIL: Test (0.04s)
panic: interface conversion: ast.Expr is *ast.CallExpr, not *ast.Ident [recovered]
	panic: interface conversion: ast.Expr is *ast.CallExpr, not *ast.Ident

goroutine 36 [running]:
testing.tRunner.func1.2({0xa0bba0, 0xc000510c30})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0xa0bba0, 0xc000510c30})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).getFuncNameFromSelector(0x0, 0xca93a0)
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/codegen.go:2072 +0x24b
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).analyzeFuncUsage.func2({0xcb0f98, 0xc000506800})
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/analysis.go:318 +0x133
go/ast.inspector.Visit(0xc000506d80, {0xcb0f98, 0xc000506800})
	/usr/local/go/src/go/ast/walk.go:375 +0x31
go/ast.Walk({0xca93a0, 0xc000506d80}, {0xcb0f98, 0xc000506800})
	/usr/local/go/src/go/ast/walk.go:50 +0x5f
go/ast.walkExprList({0xca93a0, 0xc000506d80}, {0xc000417250, 0x1, 0x0})
	/usr/local/go/src/go/ast/walk.go:24 +0x87
go/ast.Walk({0xca93a0, 0xc000506d80}, {0xcb14c0, 0xc000490740})
	/usr/local/go/src/go/ast/walk.go:217 +0x124f
go/ast.walkStmtList({0xca93a0, 0xc000506d80}, {0xc000490760, 0x2, 0x7f013019f878})
	/usr/local/go/src/go/ast/walk.go:30 +0x87
go/ast.Walk({0xca93a0, 0xc000506d80}, {0xcb0f48, 0xc0005101e0})
	/usr/local/go/src/go/ast/walk.go:225 +0xedf
go/ast.Walk({0xca93a0, 0xc000506d80}, {0xcb1218, 0xc000510210})
	/usr/local/go/src/go/ast/walk.go:346 +0x7dc
go/ast.Inspect(...)
	/usr/local/go/src/go/ast/walk.go:387
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).analyzeFuncUsage(0xc0005083c0)
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/analysis.go:311 +0x38e
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).compile(0xc0005083c0, 0xc000510630, 0x1)
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/codegen.go:2098 +0x115
github.com/nspcc-dev/neo-go/pkg/compiler.codeGen(0xc000510630)
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/codegen.go:2176 +0x353
github.com/nspcc-dev/neo-go/pkg/compiler.CompileWithOptions({0xaa88f7, 0xf72988}, {0xca8c20, 0xc000293020}, 0x0)
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/compiler.go:218 +0x65
github.com/nspcc-dev/neo-go/pkg/compiler_test.vmAndCompileInterop(0x24, {0xad9593, 0xdf})
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/vm_test.go:75 +0x113
github.com/nspcc-dev/neo-go/pkg/compiler_test.eval(0x62cc436f, {0xad9593, 0x445e37f24421}, {0x9d0b00, 0x10ac300})
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/vm_test.go:36 +0x2d
github.com/nspcc-dev/neo-go/pkg/compiler_test.Test(0x4079f9)
	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/compiler/syscall_test.go:504 +0x33
testing.tRunner(0xc00029c340, 0xc07ed0)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
```

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -918,7 +918,7 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 
 			f, ok = c.funcs[name]
 			if ok {
-				f.selector = fun.X.(*ast.Ident)
+				f.selector = fun.X
 				isBuiltin = isCustomBuiltin(f)
 				if canInline(f.pkg.Path(), f.decl.Name.Name) {
 					c.inlineCall(f, n)
@@ -2069,15 +2069,16 @@ func (c *codegen) getFuncFromIdent(fun *ast.Ident) (*funcScope, bool) {
 // getFuncNameFromSelector returns fully-qualified function name from the selector expression.
 // Second return value is true iff this was a method call, not foreign package call.
 func (c *codegen) getFuncNameFromSelector(e *ast.SelectorExpr) (string, bool) {
-	ident := e.X.(*ast.Ident)
 	if c.typeInfo.Selections[e] != nil {
-		typ := c.typeInfo.Types[ident].Type.String()
+		typ := c.typeInfo.Types[e.X].Type.String()
 		name := c.getIdentName(typ, e.Sel.Name)
 		if name[0] == '*' {
 			name = name[1:]
 		}
 		return name, true
 	}
+
+	ident := e.X.(*ast.Ident)
 	return c.getIdentName(ident.Name, e.Sel.Name), false
 }
 

--- a/pkg/compiler/func_scope.go
+++ b/pkg/compiler/func_scope.go
@@ -13,7 +13,7 @@ type funcScope struct {
 
 	// Selector of the function if there is any. Only functions imported
 	// from other packages should have a selector.
-	selector *ast.Ident
+	selector ast.Expr
 
 	// The declaration of the function in the AST. Nil if this scope is not a function.
 	decl *ast.FuncDecl

--- a/pkg/compiler/function_call_test.go
+++ b/pkg/compiler/function_call_test.go
@@ -11,6 +11,27 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestReturnValueReceiver(t *testing.T) {
+	t.Run("regular", func(t *testing.T) {
+		src := `package foo
+		import "github.com/nspcc-dev/neo-go/pkg/compiler/testdata/method"
+
+		func Main() int {
+			return method.NewX().GetA()
+		}`
+		eval(t, src, big.NewInt(42))
+	})
+	t.Run("inline", func(t *testing.T) {
+		src := `package foo
+		import "github.com/nspcc-dev/neo-go/pkg/compiler/testdata/inline"
+
+		func Main() int {
+			return inline.NewT().GetN()
+		}`
+		eval(t, src, big.NewInt(42))
+	})
+}
+
 func TestSimpleFunctionCall(t *testing.T) {
 	src := `
 		package testcase

--- a/pkg/compiler/testdata/inline/inline.go
+++ b/pkg/compiler/testdata/inline/inline.go
@@ -56,3 +56,11 @@ func (t *T) Inc(i int) int {
 	t.N += i
 	return n
 }
+
+func NewT() T {
+	return T{N: 42}
+}
+
+func (t T) GetN() int {
+	return t.N
+}

--- a/pkg/compiler/testdata/method/struct.go
+++ b/pkg/compiler/testdata/method/struct.go
@@ -1,0 +1,16 @@
+package method
+
+// X is some type.
+type X struct {
+	a int
+}
+
+// GetA returns the value of a.
+func (x X) GetA() int {
+	return x.a
+}
+
+// NewX creates a new X instance.
+func NewX() X {
+	return X{a: 42}
+}

